metaseq/scripts/generation_benchmarks.py

```diff
@@ -0,0 +1,279 @@
+from metaseq.distributed import utils as dist_utils
+from metaseq.distributed import fsdp_enable_wrap, fsdp_wrap
+from metaseq.dataclass.configs import MetaseqConfig
+import argparse
+import urllib.request
+import tarfile
+import shutil
+
+prompts = []
+
+
+def load_prompts(path, max_len=10, max_sentences=20):
+    all_data = list(open(path, "r"))
+    for data in all_data:
+        json_data = json.loads(data)
+        s = " ".join(json_data["text"].split()[:max_len])
+        prompts.append(s)
+        if len(prompts) == max_sentences:
+            break
+
+
+def setup_vocab_and_merges(model_path):
+    vocab_file = os.path.join(model_path, "gpt2-vocab.json")
+    merges_file = os.path.join(model_path, "gpt2-merges.txt")
+    tokenizer = GPT2Tokenizer(vocab_file, merges_file)
+    tokenizer.save_pretrained(model_path)
+    return vocab_file, merges_file, tokenizer
+
+
+def load_mp_model_and_run_eval(cfg: MetaseqConfig, **kwargs):
+    _, _, tokenizer = setup_vocab_and_merges(kwargs["model_path"])
+    load_prompts(kwargs["prompt_path"])
+    prompt_ids = []
+    thread_times = [
+        torch.zeros(1).cuda() for _ in range(dist_utils.get_model_parallel_world_size())
+    ]
+    begin = time.time()
+    for prompt in prompts:
+        input_ids = tensorize_input(tokenizer, prompt)
+        input_ids = F.pad(
+            input=input_ids,
+            pad=(0, kwargs["padding_size"] - input_ids.shape[1], 0, 0),
+            value=1,
+        )
+        prompt_ids.append(input_ids)
+    end = time.time()
+
+    prompt_ids = torch.cat(prompt_ids).cuda()
+    prompt_loading = end - begin
+
+    task = tasks.setup_task(cfg.task)
+
+    def _build_model(cfg, task):
+        # hardcoded to cpu & fp16
+        # Hardcoded tensor_parallel_init_model_on_gpu to be True
+        cfg.model.tensor_parallel_init_model_on_gpu = True
+        model = task.build_model(cfg.model).half().cuda()
+        return fsdp_wrap(model)
+
+    with fsdp_enable_wrap(
+        cfg.distributed_training,
+        use_sharded_state=cfg.distributed_training.use_sharded_state,
+    ):
+        models, _model_args, _task = checkpoint_utils.load_model_ensemble_and_task(
+            utils.split_paths(cfg.common_eval.path),
+            arg_overrides=None,
+            task=task,
+            suffix=cfg.checkpoint.checkpoint_suffix,
+            strict=True,
+            num_shards=cfg.checkpoint.checkpoint_shard_count,
+            build_model_hook=_build_model,
+        )
+        model = models[0]
+
+    model.summon_full_params()
+    model = model.eval()
+
+    begin = time.time()
+    with torch.no_grad():
+        model(prompt_ids)[0]
+
+    end = time.time()
+    times_list = torch.tensor(end - begin + prompt_loading).cuda()
+
+    torch.distributed.all_gather(
+        thread_times, times_list, group=dist_utils.get_global_group()
+    )
+    torch.save(thread_times, "./dependencies/times.pt")
```

Comment:
you don't need to save an external file. you can pass return values up from the distributed.call_main function.